WhateverGreen/kern_igfx.cpp

```diff
@@ -651,6 +721,585 @@ int IGFX::wrapReadAUX(void *that, IORegistryEntry *framebuffer, uint32_t address
+	
+	// Calculate the CRTC parameters
+	uint32_t multiplier = (uint32_t) (context.frequency / 24000000);
+	uint32_t fraction = (uint32_t) (context.frequency - multiplier * 24000000);
+	uint32_t cf15625 = (uint32_t) (context.central / 15625);
+	DBGLOG("igfx", "SC: ComputeHdmiP0P1P2() DInfo: Multiplier = %d; Fraction = %d; CF15625 = %d.\n", multiplier, fraction, cf15625);
+	// Guard: The given CRTC parameters should never be NULL
+	auto params = reinterpret_cast<CRTCParams*>(parameters);
+	if (params == nullptr) {
+		DBGLOG("igfx", "SC: ComputeHdmiP0P1P2() Error: The given CRTC parameters should not be NULL.");
+		return 0;
+	}
+	params->pdiv = context.pdiv;
+	params->qdiv = context.qdiv;
+	params->kdiv = context.kdiv;
+	params->multiplier = multiplier;
+	params->fraction = fraction;
+	params->cf15625 = cf15625;
+	DBGLOG("igfx", "SC: ComputeHdmiP0P1P2() DInfo: CTRC parameters have been populated successfully.");
+	return 0;
+}
+
+IGFX::LSPCON::LSPCON(void *controller, IORegistryEntry *framebuffer, void *displayPath) {
+	this->controller = controller;
+	this->framebuffer = framebuffer;
+	this->displayPath = displayPath;
+	this->isActive = false;
+	this->index = OSDynamicCast(OSNumber, framebuffer->getProperty("IOFBDependentIndex"))->unsigned32BitValue();
+}
+
+IOReturn IGFX::LSPCON::probe() {
+	// Read the adapter info
+	uint8_t buffer[128];
+	bzero(buffer, 128);
+	IOReturn retVal = callbackIGFX->advReadI2COverAUX(controller, framebuffer, displayPath, DP_DUAL_MODE_ADAPTER_I2C_ADDR, 0x00, 128, buffer, 0);
+	if (retVal != kIOReturnSuccess) {
+		SYSLOG("igfx", "SC: LSPCON::probe() Error: [FB%d] Failed to read the LSPCON adapter info. RV = 0x%llx.", index, retVal);
+		return retVal;
+	}
+	
+	// Start to parse the adapter info
+	auto info = reinterpret_cast<DisplayPortDualModeAdapterInfo*>(buffer);
+	// Guard: Check whether this is a LSPCON adapter
+	if (!isLSPCONAdapter(info)) {
+		SYSLOG("igfx", "SC: LSPCON::probe() Error: [FB%d] Not a LSPCON DP-HDMI adapter. AdapterID = 0x%02x.", index, info->adapterID);
+		return kIOReturnNotFound;
+	}
+	
+	// Parse the chip vendor
+	char device[8];
+	bzero(device, 8);
+	lilu_os_memcpy(device, info->deviceID, 6);
+	DBGLOG("igfx", "SC: LSPCON::probe() DInfo: [FB%d] Found the LSPCON adapter: %s %s.", index, getVendorString(parseVendor(info)), device);
+	
+	// Parse the current adapter mode
+	Mode mode = parseMode(info->lspconCurrentMode);
+	DBGLOG("igfx", "SC: LSPCON::probe() DInfo: [FB%d] The current adapter mode is %s.", index, getModeString(mode));
+	switch (mode) {
+		case Mode::LevelShifter:
+			break;
+			
+		case Mode::ProtocolConverter:
+			isActive = true;
+			break;
+			
+		default:
+			SYSLOG("igfx", "SC: LSPCON::probe() Error: [FB%d] Cannot detect the current adapter mode. Assuming Level Shifter mode.", index);
+			break;
+	}
+	
+	return kIOReturnSuccess;
+}
+
+IOReturn IGFX::LSPCON::getMode(Mode *mode) {
+	uint8_t byte;
```

Comment:
Maybe move it right above `retVal = callbackIGFX->advReadI2COverAUX`? It is very confusing where it belongs to be honest, especially with the name. I think it may be reasonable to also rename it to hardwareMode or something like that.